File: src/hindi/vocabulary.ts
import {assert, flatten} from '../lib/base';
import {Tense} from '../parsing/extensions';

interface Entry {
  head: string;
  hindi: string;
  latin: string;
  scores: Map<string, number>;
  tenses: Tense[];
  value: string;
}

// Boilerplate helpers needed to support the core declension routines.

interface Case extends Word {
  tense: Tense;
}

interface Word {
  hindi: string;
  latin: string;
}

const kFeminine: Tense = {gender: 'feminine'};
const kMasculine: Tense = {gender: 'masculine'};

const kPlural: Tense = {count: 'plural'};
const kSingular: Tense = {count: 'singular'};

const kVerbTenses = [
  {...kMasculine, ...kSingular},
  {...kMasculine, ...kPlural},
  {...kFeminine},
];

const option = <T>(x: string, table: {[x: string]: T}, type: string): T => {
  if (x in table) return table[x];
  const keys = Object.keys(table)
    .map(x => JSON.stringify(x))
    .join(', ');
  throw new Error(`${type} must be in {${keys}}; got: ${x}`);
};

const rollup = (cases: Case[], type: string, value: string): Entry[] => {
  // Collect the set of distinct words and a base entry common to them all.
  assert(cases.length > 0, () => `Invalid cases for type: ${type}`);
  const head = `${type}-${cases[0].hindi}`;
  const hindis = Array.from(new Set(cases.map(x => x.hindi)).keys());

  // Return one entry for each distinct Hindi word.
  return hindis.map(hindi => {
    const matched_cases = cases.filter(x => x.hindi === hindi);
    const latin = matched_cases[0].latin;
    const scores = new Map([[`%${type}`, 0]]);
    cases.forEach(y => scores.set(y.latin, y.hindi === hindi ? 0 : -1));
    const tenses = matched_cases.map(x => x.tense);
    return {head, hindi, latin, scores, tenses, value};
  });
};

const split = (x: string): Word => {
  const result = x.split('/');
  if (result.length !== 2) throw Error(`Invalid word: ${x}`);
  return {hindi: result[1], latin: result[0]};
};

const zip = (hindis: string[], latins: string[], tenses: Tense[]): Case[] => {
  assert(hindis.length === latins.length && latins.length === tenses.length);
  tenses = tenses.map(x => ({...x}));
  return hindis.map((x, i) => ({hindi: x, latin: latins[i], tense: tenses[i]}));
};

// The table parsing code follows.

const parse_table = <T>(t: T, table: string): {[k in keyof T]: string}[] => {
  // Split the table into lines and validate the header.
  const lines = table.split('\n').map(x => x.trim());
  const valid = lines.filter(x => x.length !== 0 && x[0] !== '#');
  if (valid.length < 3) throw Error(`Invalid table:\n${table}`);
  const columns = valid[0].split('|').map(x => x.trim());
  const keys = Object.keys(t);
  if (keys.length !== columns.length || keys.some(x => !columns.includes(x))) {
    const expected = `(expected: ${keys.join(', ')})`;
    throw Error(`Invalid header: ${columns.join(', ')} ${expected}`);
  }

  // Return a value for each entry of the actual table.
  const grid = valid.slice(2).map(x => x.split('|').map(y => y.trim()));
  return grid.map((x, i) => {
    if (x.length !== columns.length) {
      throw Error(`Invalid line: ${valid[i + 2]}`);
    }
    // tslint:disable-next-line:no-any
    const result: any = {};
    for (let j = 0; j < x.length; j++) {
      if (x[j] === '<') x[j] = x[j - 1];
      if (x[j] === '^') x[j] = grid[i - 1][j];
      if (!x[j]) throw Error(`Invalid line: ${valid[i + 2]}`);
      result[columns[j]] = x[j];
    }
    return result;
  });
};

// Helpers for automatically computing different forms of words.

const pluralize = (gender: string, noun: Word): Word => {
  const {hindi, latin} = noun;
  if (gender === 'm' && hindi.endsWith('A')) {
    return {hindi: `${hindi.slice(0, -1)}e`, latin: `${latin.slice(0, -1)}e`};
  } else if (gender === 'f' && hindi.endsWith('I')) {
    return {hindi: `${hindi}yAM`, latin: `${latin}ya`};
  }
  const kOverrides: {[spec: string]: string} = {
    'aurat/Oraw': 'aurte/Orwe',
  };
  const spec = kOverrides[`${latin}/${hindi}`];
  if (!spec) throw Error(`Unable to pluralize: ${hindi}`);
  return split(spec);
};

// Helpers for generating declined adjectives, nouns, and verbs.

const adjectives = (table: string): Entry[] => {
  const entries: Entry[][] = [];
  const example = {meaning: 'small', word: 'chota/cotA'};
  for (const {meaning, word} of parse_table(example, table)) {
    const {hindi, latin} = split(word);
    if (hindi.endsWith('A')) {
      const hindis = ['A', 'e', 'I'].map(x => `${hindi.slice(0, -1)}${x}`);
      const latins = ['a', 'e', 'i'].map(x => `${latin.slice(0, -1)}${x}`);
      const tenses = kVerbTenses;
      entries.push(rollup(zip(hindis, latins, tenses), 'adjective', meaning));
    } else {
      entries.push(rollup([{hindi, latin, tense: {}}], 'adjective', meaning));
    }
  }
  return flatten(entries);
};

const nouns = (table: string): Entry[] => {
  const entries: Entry[][] = [];
  const t = {category: 0, role: 0, meaning: 0, word: 0};
  for (const row of parse_table(t, table)) {
    const {category, role, meaning, word} = row;
    const gender = option(role[0], {f: kFeminine, m: kMasculine}, 'gender');
    const {hindi, latin} = split(word);
    const tense = {...gender, person: 'third'};

    // Create multiple forms for nouns with a plural declension.
    if (role[1] !== '.') {
      const singular = {hindi, latin};
      const splits = [singular, pluralize(role[0], singular)];
      const hindis = splits.map(x => x.hindi);
      const latins = splits.map(x => x.latin);
      const tenses = [kSingular, kPlural].map(x => ({...tense, ...x}));
      entries.push(rollup(zip(hindis, latins, tenses), 'noun', meaning));
    } else {
      entries.push(rollup([{hindi, latin, tense}], 'noun', meaning));
    }

    // Add types to each entry based on the category and the count.
    const xs = entries[entries.length - 1];
    xs.forEach(x => {
      x.scores.set(`%${category}`, 0);
      let counts = x.tenses.filter(x => !!x.count);
      if (counts.length === 0) counts = [kSingular, kPlural];
      counts.forEach(y => x.scores.set(`%noun_${y.count}`, 0));
      counts.forEach(y => x.scores.set(`%${category}_${y.count}`, 0));
    });
  }
  return flatten(entries);
};

const numbers = (table: string): Entry[] => {
  const entries: Entry[][] = [];
  const example = {meaning: '0', word: 'sifar/siPZar'};
  for (const {meaning, word} of parse_table(example, table)) {
    const value = parseInt(meaning, 10);
    if (isNaN(value)) throw Error(`Invalid number: ${value}`);
    const {hindi, latin} = split(word);
    const tense = value === 1 ? kSingular : kPlural;
    entries.push(rollup([{latin, hindi, tense}], 'number', meaning));
  }
  return flatten(entries);
};

const particles = (table: string): Entry[] => {
  const entries: Entry[][] = [];
  const t = {category: 0, meaning: 0, word: 0, declines: 0};
  for (const row of parse_table(t, table)) {
    const {category, declines, meaning, word} = row;
    const {hindi, latin} = split(word);
    if (option(declines, {n: false, y: true}, 'declines')) {
      if (!hindi.endsWith('A')) throw Error(`Unable to decline: ${hindi}`);
      const hindis = ['A', 'e', 'I'].map(x => `${hindi.slice(0, -1)}${x}`);
      const latins = ['a', 'e', 'i'].map(x => `${latin.slice(0, -1)}${x}`);
      const tenses = kVerbTenses;
      entries.push(rollup(zip(hindis, latins, tenses), 'particle', meaning));
    } else {
      entries.push(rollup([{hindi, latin, tense: {}}], 'particle', meaning));
    }
    entries[entries.length - 1].forEach(x => x.scores.set(`%${category}`, 0));
  }
  return flatten(entries);
};

const pronouns = (table: string): Entry[] => {
  const entries: Entry[][] = [];
  const d = {'dative (1)': '', 'dative (2)': ''};
  const t = {role: '', direct: '', genitive: '', copula: '', ...d};

  const persons = {1: 'first', 2: 'second', 3: 'third'};
  const counts = {p: 'plural', s: 'singular'};
  const tones = {c: 'casual', f: 'formal', i: 'intimate', ...{'.': null}};
  const meaning = {first: 'I', second: 'you', third: 'they'};

  const group_by_person: {[key: string]: (typeof t)[]} = {};
  for (const row of parse_table({...d, ...t}, table)) {
    const person = option(row.role[0], persons, 'person');
    (group_by_person[person] = group_by_person[person] || []).push(row);
  }

  const copulas: Case[] = [];
  const fn = (): Case[][] => [];

  for (const person of Object.keys(group_by_person).sort()) {
    const cases = {dative: fn(), direct: fn(), genitive: fn()};
    const value = option(person, meaning, 'meaning');

    for (const row of group_by_person[person]) {
      const {copula, direct, genitive, role} = row;
      if (role.length !== 3) throw Error(`Invalid role: ${role}`);
      const count = option(role[1], counts, 'count');
      const tone = option(role[2], tones, 'tone');
      const base = tone ? {tone} : null;
      const tense = {person, count, ...base};

      {
        // Handle the direct (subject) case.
        const {hindi, latin} = split(direct);
        cases.direct.push([{hindi, latin, tense}]);
      }
      {
        // Handle the genitive (possessive) case.
        const {hindi, latin} = split(genitive);
        if (!hindi.endsWith('A')) throw Error(`Unable to decline: ${hindi}`);
        const hindis = ['A', 'e', 'I'].map(x => `${hindi.slice(0, -1)}${x}`);
        const latins = ['a', 'e', 'i'].map(x => `${latin.slice(0, -1)}${x}`);
        const tenses = kVerbTenses.map(x => ({...base, ...x}));
        cases.genitive.push(zip(hindis, latins, tenses));
      }
      {
        // Handle the dative (indirect object) case.
        const datives = new Set([row['dative (1)'], row['dative (2)']]);
        const splits = Array.from(datives).map(split);
        const hindis = splits.map(x => x.hindi);
        const latins = splits.map(x => x.latin);
        const tenses = splits.map(x => ({...base}));
        cases.dative.push(zip(hindis, latins, tenses));
      }
      {
        // Handle the copula case. The copulas will all have a single stem.
        const {hindi, latin} = split(copula);
        copulas.push({hindi, latin, tense});
      }
    }

    entries.push(rollup(flatten(cases.dative), 'dative', value));
    entries.push(rollup(flatten(cases.direct), 'direct', value));
    entries.push(rollup(flatten(cases.genitive), 'genitive', value));
  }

  entries.push(rollup(copulas, 'copula', 'be'));
  return flatten(entries);
};

const verbs = (table: string): Entry[] => {
  // TODO(skishore): Add command forms here.
  // TODO(skishore): Handle "reversed" verbs like "chahna".
  // TODO(skishore): Handle irregular verbs here ("hona", "jana", etc.)
  // I can't find a complete list. http://www.geocities.ws/lordvaruna/
  // notes all the irregular verbs, but doesn't show some of the modified
  // conjugutions (e.g. the changed future tenses).
  const entries: Entry[][] = [];
  const kBases = [
    {hindi: '', latin: '', type: 'stem'},
    {hindi: 'ne', latin: 'ne', type: 'gerund'},
    {hindi: 'nA', latin: 'na', type: 'infinitive'},
  ];
  const kForms = [
    {hindi: '', latin: '', prefix: true, time: 'past'},
    {hindi: 'w', latin: 't', prefix: false, time: 'present'},
  ];
  const example = {meaning: 'eat', word: 'khana/KAnA'};
  for (const {meaning, word} of parse_table(example, table)) {
    // Detect the stem and check if it ends in a vowel.
    const {hindi, latin} = split(word);
    assert(
      hindi.endsWith('nA') && latin.endsWith('na'),
      () => `Invalid infinitive: ${latin}`,
    );
    const h = hindi.slice(0, -2);
    const l = latin.slice(0, -2);
    const vowel = 'aeiou'.indexOf(h.slice(-1).toLowerCase()) >= 0;

    // For each base type, add an entry for the verb.
    for (const base of kBases) {
      const hindis = [`${h}${base.hindi}`];
      const latins = [`${l}${base.latin}`];
      entries.push(rollup(zip(hindis, latins, [{}]), 'verb', meaning));
      const xs = entries[entries.length - 1];
      xs.forEach(x => x.scores.set(`%verb_${base.type}`, 0));
    }

    // For each form type, add declined entries for that type.
    for (const form of kForms) {
      const y = vowel && form.prefix ? 'y' : '';
      const hindis = ['A', 'e', 'I'].map(x => `${h}${y}${form.hindi}${x}`);
      const latins = ['a', 'e', 'i'].map(x => `${l}${y}${form.latin}${x}`);
      const tenses = kVerbTenses.map(x => ({...x, time: form.time}));
      entries.push(rollup(zip(hindis, latins, tenses), 'verb', meaning));
      const xs = entries[entries.length - 1];
      xs.forEach(x => x.scores.set(`%verb_${form.time}`, 0));
    }

    // The future tense is special: it has different forms based on person.
    for (const form of [{time: 'future'}]) {
      const hs = 'UngA egA egA enge ogA enge enge'.split(' ');
      const ls = 'unga ega ega enge oga enge enge'.split(' ');
      const basics = [
        {...kSingular, person: 'first'},
        {...kSingular, person: 'second'},
        {...kSingular, person: 'third'},
        {...kPlural, person: 'first'},
        {...kPlural, person: 'second', tone: 'casual'},
        {...kPlural, person: 'second', tone: 'formal'},
        {...kPlural, person: 'third'},
      ];
      const hindis = hs
        .concat(hs.map(x => `${x.slice(0, -1)}I`))
        .map(x => `${h}${x}`);
      const latins = ls
        .concat(ls.map(x => `${x.slice(0, -1)}i`))
        .map(x => `${l}${x}`);
      const tenses = basics
        .map(x => ({...x, ...kMasculine, time: 'future'}))
        .concat(basics.map(x => ({...x, ...kFeminine, time: 'future'})));
      entries.push(rollup(zip(hindis, latins, tenses), 'verb', meaning));
      const xs = entries[entries.length - 1];
      xs.forEach(x => x.scores.set(`%verb_${form.time}`, 0));
    }
  }
  return flatten(entries);
};

const Vocabulary = {adjectives, nouns, numbers, particles, pronouns, verbs};

export {Entry, Vocabulary};

File: src/hindi/vocabulary.rs
use super::super::lib::base::{HashMap, Result};
use super::super::nlu::base::Tense;
use super::wx::wx_to_hindi;

pub struct Entry {
  pub head: String,
  pub hindi: String,
  pub latin: String,
  pub scores: HashMap<String, f32>,
  pub tenses: Vec<Tense>,
  pub value: String,
}

thread_local! {
  static CATEGORIES: Vec<(&'static str, Vec<(u8, &'static str)>)> = vec![
    ("count", vec![(b'p', "plural"), (b's', "singular")]),
    ("gender", vec![(b'f', "female"), (b'm', "male")]),
    ("person", vec![(b'1', "first"), (b'2', "second"), (b'3', "third")]),
    ("time", vec![(b'<', "past"), (b'=', "present"), (b'>', "future")]),
    ("tone", vec![(b'c', "casual"), (b'f', "formal"), (b'i', "intimate")]),
  ];
}

// Some helpers. Call rollup to construct a list of related vocabulary result. Call tense
// to build a Tense object with interned (statically-allocated) string keys and values.

struct Case {
  hindi: String,
  latin: String,
  tense: Tense,
}

fn rollup(cases: &[Case], class: &str, value: &str) -> Result<Vec<Entry>> {
  assert!(!cases.is_empty());
  let head = format!("{}-{}", class, cases[0].hindi);
  let mut result: Vec<Entry> = vec![];

  for case in cases {
    if result.iter().any(|x| x.hindi == case.hindi) {
      continue;
    } else if case.latin != case.latin.to_lowercase() {
      Err(format!("Invalid Latin: {}", case.latin))?;
    }
    wx_to_hindi(&case.hindi)?;
    let (hindi, latin) = (case.hindi.clone(), case.latin.clone());
    let scores = std::iter::once((format!("%{}", class), 0.0))
      .chain(cases.iter().map(|x| (x.latin.clone(), if x.hindi == hindi { 0.0 } else { -1.0 })))
      .collect();
    let tenses = cases.iter().filter(|x| x.hindi == hindi).map(|x| x.tense.clone()).collect();
    result.push(Entry { head: head.clone(), hindi, latin, scores, tenses, value: value.into() });
  }
  Ok(result)
}

fn split(word: &str) -> Result<(String, String)> {
  let index = word.find('/').ok_or_else(|| format!("Invalid word (missing slash): {}", word))?;
  let (hindi, latin) = (&word[index + 1..], &word[..index]);
  Ok((hindi.to_string(), latin.to_string()))
}

fn tense(code: &str) -> Result<Tense> {
  CATEGORIES.with(|categories| {
    if code.len() != categories.len() {
      Err(format!("Invalid tense code: {}", code))?
    }
    let mut result = HashMap::default();
    for (i, ch) in code.as_bytes().iter().cloned().enumerate().filter(|x| x.1 != b'.') {
      let (category, values) = &categories[i];
      let maybe = values.iter().find(|x| x.0 == ch);
      let value = maybe.ok_or_else(|| format!("Invalid tense code: {}", code))?;
      result.insert(*category, value.1);
    }
    Tense::new(&result)
  })
}

fn zip(hindis: Vec<String>, latins: Vec<String>, tenses: Vec<Tense>) -> Vec<Case> {
  assert!(hindis.len() == latins.len() && latins.len() == tenses.len());
  let iter = hindis.into_iter().zip(latins.into_iter()).zip(tenses.into_iter());
  iter.map(|x| Case { hindi: (x.0).0, latin: (x.0).1, tense: x.1 }).collect()
}

// Our public interface is a series of functions that can be used to build vocabulary result.

pub fn adjectives(table: &str) -> Result<Vec<Entry>> {
  let mut result = vec![];
  for_each_row!(table, [meaning, word], {
    let (hindi, latin) = split(word)?;
    if hindi.ends_with('A') && latin.ends_with('a') {
      let (hstem, lstem) = (&hindi[..hindi.len() - 1], &latin[..latin.len() - 1]);
      let hindis: Vec<_> = ['A', 'e', 'I'].iter().map(|x| format!("{}{}", hstem, x)).collect();
      let latins: Vec<_> = ['a', 'e', 'i'].iter().map(|x| format!("{}{}", lstem, x)).collect();
      let tenses: Vec<_> = ["sm...", "pm...", ".f..."].iter().map(|x| tense(x).unwrap()).collect();
      result.push(rollup(&zip(hindis, latins, tenses), "adjective", meaning)?);
    } else {
      result.push(rollup(&[Case { hindi, latin, tense: Tense::default() }], "adjective", meaning)?);
    }
  });
  Ok(result.into_iter().flatten().collect())
}

pub fn nouns(main: &str, supplement: &str) -> Result<Vec<Entry>> {
  let mut plurals = HashMap::default();
  for_each_row!(supplement, [singular, plural], {
    plurals.insert(singular, plural);
  });
  let mut result = vec![];
  let default_counts = vec!["singular".to_string(), "plural".to_string()];
  for_each_row!(main, [category, meaning, word, role], {
    let (hindi, latin) = split(word)?;
    let (gender, declines) = match role {
      "m." => ('m', false),
      "f." => ('f', false),
      "ms" => ('m', true),
      "fs" => ('f', true),
      _ => Err(format!("Invalid noun role: {}", role))?,
    };

    // Create singular and plural forms for nouns that decline.
    if declines {
      let (hp, lp) = plurals.remove(word).map(split).unwrap_or_else(|| {
        if gender == 'm' && hindi.ends_with('A') && latin.ends_with('a') {
          let (hstem, lstem) = (&hindi[..hindi.len() - 1], &latin[..latin.len() - 1]);
          return Ok((format!("{}e", hstem), format!("{}e", lstem)));
        } else if gender == 'f' && hindi.ends_with('I') && latin.ends_with('i') {
          return Ok((format!("{}yAM", hindi), format!("{}ya", latin)));
        }
        Err(format!("Unable to pluralize noun: {}", word))?
      })?;
      let tenses = vec![tense(&format!("s{}3..", gender))?, tense(&format!("p{}3..", gender))?];
      result.push(rollup(&zip(vec![hindi, hp], vec![latin, lp], tenses), "noun", meaning)?);
    } else {
      let tense = tense(&format!(".{}3..", gender))?;
      result.push(rollup(&[Case { hindi, latin, tense }], "noun", meaning)?);
    }

    // Add types to each entry based on the category and the count.
    let last = result.last_mut().unwrap();
    last.iter_mut().for_each(|x| {
      x.scores.insert(format!("%{}", category), 0.0);
      let counts: Vec<_> = x.tenses.iter().filter_map(|y| y.get("count")).collect();
      let nonempty = if counts.is_empty() { &default_counts } else { &counts };
      nonempty.iter().for_each(|y| {
        x.scores.insert(format!("%noun_{}", y), 0.0);
        x.scores.insert(format!("%{}_{}", category, y), 0.0);
      });
    });
  });

  if !plurals.is_empty() {
    let unused: Vec<_> = plurals.into_iter().map(|x| x.0.to_string()).collect();
    Err(format!("Unused plural nouns: {}", unused.join(", ")))?
  }
  Ok(result.into_iter().flatten().collect())
}

pub fn numbers(table: &str) -> Result<Vec<Entry>> {
  let mut result = vec![];
  for_each_row!(table, [meaning, word], {
    let value = meaning.parse::<usize>().map_err(|_| format!("Invalid number: {}", meaning))?;
    let (hindi, latin) = split(word)?;
    let tense = tense(if value == 1 { "s...." } else { "p...." }).unwrap();
    result.push(rollup(&[Case { hindi, latin, tense }], "number", meaning)?);
  });
  Ok(result.into_iter().flatten().collect())
}

pub fn particles(table: &str) -> Result<Vec<Entry>> {
  let mut result = vec![];
  for_each_row!(table, [category, meaning, word, declines], {
    let (hindi, latin) = split(word)?;
    let declines = match declines {
      "n" => false,
      "y" => true,
      _ => Err(format!("declines must be n or y; got: {}", declines))?,
    };

    // Create male singular, male plural, and female forms for particles that decline.
    if declines {
      if !(hindi.ends_with('A') && latin.ends_with('a')) {
        Err(format!("Declining particles must end in A. Got: {}", word))?
      }
      let (hstem, lstem) = (&hindi[..hindi.len() - 1], &latin[..latin.len() - 1]);
      let hindis: Vec<_> = ['A', 'e', 'I'].iter().map(|x| format!("{}{}", hstem, x)).collect();
      let latins: Vec<_> = ['a', 'e', 'i'].iter().map(|x| format!("{}{}", lstem, x)).collect();
      let tenses: Vec<_> = ["sm...", "pm...", ".f..."].iter().map(|x| tense(x).unwrap()).collect();
      result.push(rollup(&zip(hindis, latins, tenses), "particle", meaning)?);
    } else {
      result.push(rollup(&[Case { hindi, latin, tense: Tense::default() }], "particle", meaning)?);
    }

    // Add types to particles based on their category.
    let last = result.last_mut().unwrap();
    last.iter_mut().for_each(|x| std::mem::drop(x.scores.insert(format!("%{}", category), 0.0)));
  });
  Ok(result.into_iter().flatten().collect())
}

pub fn pronouns(table: &str) -> Result<Vec<Entry>> {
  let mut groups = HashMap::default();
  for_each_row!(table, [role, direct, genitive, dative_1, dative_2, copula], {
    if !(role.len() == 3 && role.is_ascii() && role.find(|c| ('1'..='3').contains(&c)) == Some(0)) {
      Err(format!("Invalid pronoun role: {}", role))?
    }
    let (person, count, tone) = (&role[..1], &role[1..2], &role[2..]);
    let basis = tense(&format!("{}.{}.{}", count, person, tone))?;
    let entry = (basis, copula, dative_1, dative_2, direct, genitive);
    groups.entry(person).or_insert(vec![]).push(entry);
  });
  let (mut copula_cases, mut result) = (vec![], vec![]);
  for (person, value) in &[("1", "I"), ("2", "you"), ("3", "they")] {
    for entry in groups.get(person).unwrap_or(&vec![]) {
      let (basis, copula, dative_1, dative_2, direct, genitive) = entry;
      copula_cases.push({
        let (hindi, latin) = split(copula)?;
        Case { hindi, latin, tense: basis.clone() }
      });
      let direct_cases = {
        let (hindi, latin) = split(direct)?;
        vec![Case { hindi, latin, tense: basis.clone() }]
      };
      let genitive_cases = {
        let (hindi, latin) = split(genitive)?;
        if !(hindi.ends_with('A') && latin.ends_with('a')) {
          Err(format!("Genitive pronouns must end in A. Got: {}", genitive))?
        }
        let (hstem, lstem) = (&hindi[..hindi.len() - 1], &latin[..latin.len() - 1]);
        let hindis: Vec<_> = ['A', 'e', 'I'].iter().map(|x| format!("{}{}", hstem, x)).collect();
        let latins: Vec<_> = ['a', 'e', 'i'].iter().map(|x| format!("{}{}", lstem, x)).collect();
        let ts: Vec<_> = ["sm...", "pm...", ".f..."].iter().map(|x| tense(x).unwrap()).collect();
        zip(hindis, latins, ts)
      };
      let dative_cases = {
        let datives = if dative_1 == dative_2 { vec![dative_1] } else { vec![dative_1, dative_2] };
        let splits = datives.into_iter().map(|x| split(x)).collect::<Result<Vec<_>>>()?;
        let hindis: Vec<_> = splits.iter().map(|x| x.0.to_string()).collect();
        let latins: Vec<_> = splits.iter().map(|x| x.1.to_string()).collect();
        let tenses: Vec<_> = splits.iter().map(|_| basis.clone()).collect();
        zip(hindis, latins, tenses)
      };
      result.push(rollup(&dative_cases, "dative", value)?);
      result.push(rollup(&direct_cases, "direct", value)?);
      result.push(rollup(&genitive_cases, "genitive", value)?);
    }
  }
  result.push(rollup(&copula_cases, "copula", "be")?);
  Ok(result.into_iter().flatten().collect())
}

pub fn verbs(table: &str) -> Result<Vec<Entry>> {
  // TODO(skishore): Add command forms here.
  // TODO(skishore): Handle "reversed" verbs like "chahna".
  // TODO(skishore): Handle irregular verbs here ("hona", "jana", etc.)
  let mut result = vec![];
  let base_forms = [("", "", "stem"), ("ne", "ne", "gerund"), ("nA", "na", "infinitive")];
  let time_forms = [("", "", "past", true), ("w", "t", "present", false)];
  let (male, female) = (tense(".m...").unwrap(), tense(".f...").unwrap());

  for_each_row!(table, [meaning, word], {
    let (hindi, latin) = split(word)?;
    if !(hindi.ends_with('A') && latin.ends_with('a')) {
      Err(format!("Verbs must end in nA. Got: {}", word))?
    }
    let (hstem, lstem) = (&hindi[..hindi.len() - 2], &latin[..latin.len() - 2]);
    let vowel = "aeiou".chars().any(|x| hstem.to_lowercase().ends_with(x));

    // For each base type, add an entry for the verb.
    for (h, l, t) in &base_forms {
      let hindi = format!("{}{}", hstem, h);
      let latin = format!("{}{}", lstem, l);
      result.push(rollup(&[Case { hindi, latin, tense: Tense::default() }], "verb", meaning)?);
      let last = result.last_mut().unwrap();
      last.iter_mut().for_each(|x| std::mem::drop(x.scores.insert(format!("%verb_{}", t), 0.0)));
    }

    // For each temporal type, add declined entries for the verb.
    for (h, l, time, prefix) in &time_forms {
      let base = Tense::new(&vec![("time", *time)].into_iter().collect()).unwrap();
      let y = if vowel && *prefix { "y" } else { "" };
      let h: Vec<_> = ['A', 'e', 'I'].iter().map(|x| format!("{}{}{}{}", hstem, h, y, x)).collect();
      let l: Vec<_> = ['a', 'e', 'i'].iter().map(|x| format!("{}{}{}{}", lstem, l, y, x)).collect();
      let mut t: Vec<_> = ["sm...", "pm...", ".f..."].iter().map(|x| tense(x).unwrap()).collect();
      t.iter_mut().for_each(|x| x.union(&base));
      result.push(rollup(&zip(h, l, t), "verb", meaning)?);
      let last = result.last_mut().unwrap();
      last.iter_mut().for_each(|x| std::mem::drop(x.scores.insert(format!("%verb_{}", time), 0.0)));
    }

    // The future tense is special: it has different forms based on person.
    {
      let time = &"future";
      let hs: Vec<_> = "UngA  egA   egA   enge  ogA   enge  enge".split_whitespace().collect();
      let ls: Vec<_> = "unga  ega   ega   enge  oga   enge  enge".split_whitespace().collect();
      let ts: Vec<_> = "s.1.. s.2.i s.3.. p.1.. p.2.c p.2.f p.3..".split_whitespace().collect();
      let hindis = hs.iter().map(|x| x.to_string());
      let hindis = hindis
        .chain(hs.iter().map(|x| format!("{}I", &x[..x.len() - 1])))
        .map(|x| format!("{}{}", hstem, x));
      let latins = ls.iter().map(|x| x.to_string());
      let latins = latins
        .chain(ls.iter().map(|x| format!("{}i", &x[..x.len() - 1])))
        .map(|x| format!("{}{}", lstem, x));
      let tenses = ts.iter().map(|x| tense(x)).collect::<Result<Vec<_>>>()?;
      let (mut m, mut f) = (tenses.clone(), tenses);
      m.iter_mut().for_each(|x| x.union(&male));
      f.iter_mut().for_each(|x| x.union(&female));
      let tenses = m.into_iter().chain(f.into_iter()).collect();
      result.push(rollup(&zip(hindis.collect(), latins.collect(), tenses), "verb", meaning)?);
      let last = result.last_mut().unwrap();
      last.iter_mut().for_each(|x| std::mem::drop(x.scores.insert(format!("%verb_{}", time), 0.0)));
    }
  });
  Ok(result.into_iter().flatten().collect())
}

// Our overall entry point calls each of the helpers above.

pub fn vocabulary(text: &str) -> Result<Vec<Entry>> {
  let mut entries = vec![];
  let (a, b, c, d, e, f) = (adjectives, nouns, numbers, particles, pronouns, verbs);
  for_each_table!(text, [adjectives, nouns, noun_plurals, numbers, particles, pronouns, verbs], {
    entries.extend(a(adjectives)?.into_iter());
    entries.extend(b(nouns, noun_plurals)?.into_iter());
    entries.extend(c(numbers)?.into_iter());
    entries.extend(d(particles)?.into_iter());
    entries.extend(e(pronouns)?.into_iter());
    entries.extend(f(verbs)?.into_iter());
  });
  Ok(entries)
}

#[cfg(test)]
mod test {
  use super::*;

  #[test]
  fn test_all_vocabulary_entries() {
    let file = "src/hindi/hindi.grammar";
    let data = std::fs::read_to_string(file).unwrap();
    let base = regex::Regex::new(r#"lexer: ```[\s\S]*```"#).unwrap().find(&data).unwrap();
    let text = &data[base.start() + 10..base.end() - 3];
    vocabulary(text).unwrap();
  }
}
